File: frontend/src/narration-connection.ts
/**
 * NarrationConnection — Second Gemini Live session for narrating Claude's activity.
 *
 * Speaks real-time commentary while Claude is working, so the user isn't
 * left in silence. Coordinates with main Gemini: only plays audio when
 * main Gemini is waiting for a function response.
 */

import { GoogleGenAI, Modality } from "@google/genai";
import type { AudioManager } from "./audio-manager";
import type { ServerConfig, TokenResponse } from "./types";
import { log } from "./debug-log";

export class NarrationConnection {
  private session: any = null;
  private audioManager: AudioManager;
  private onTranscript: (text: string) => void;
  private muted = false;
  private connected = false;
  private languageCode: string;

  // Batch events: collect for a short window, then send as one message
  private eventBuffer: string[] = [];
  private flushTimer: ReturnType<typeof setTimeout> | null = null;
  private readonly FLUSH_DELAY_MS = 1500;

  constructor(audioManager: AudioManager, onTranscript: (text: string) => void, languageCode: string = "en-US") {
    this.audioManager = audioManager;
    this.onTranscript = onTranscript;
    this.languageCode = languageCode;
  }

  async connect(): Promise<void> {
    try {
      const [tokenRes, configRes] = await Promise.all([
        fetch("/api/token").then((r) => r.json()) as Promise<TokenResponse>,
        fetch("/api/narration-config").then((r) => r.json()) as Promise<ServerConfig>,
      ]);

      const ai = new GoogleGenAI({
        apiKey: tokenRes.token,
        httpOptions: { apiVersion: "v1alpha" },
      });

      log("NARRATION", `Connecting model=${configRes.model}`);

      this.session = await ai.live.connect({
        model: configRes.model,
        config: {
          responseModalities: [Modality.AUDIO],
          systemInstruction: configRes.system_prompt,
          speechConfig: {
            languageCode: this.languageCode,
          },
          outputAudioTranscription: {},
          realtimeInputConfig: {
            automaticActivityDetection: { disabled: true },
          },
          maxOutputTokens: 8192,
          thinkingConfig: {
            thinkingBudget: 8192,
          },
          contextWindowCompression: {
            triggerTokens: "100000",
            slidingWindow: { targetTokens: "80000" },
          },
        },
        callbacks: {
          onopen: () => {
            log("NARRATION", "Connected");
            this.connected = true;
          },
          onmessage: (message: any) => {
            this.handleMessage(message);
          },
          onerror: (error: any) => {
            log("NARRATION", "Error", error?.message || error);
          },
          onclose: (event: any) => {
            log("NARRATION", "Disconnected", `code=${event?.code} reason=${event?.reason || "unknown"}`);
            this.connected = false;
          },
        },
      });
    } catch (err) {
      console.error("Narration connection failed:", err);
    }
  }

  private handleMessage(message: any): void {
    if (this.muted) return;

    // Audio output — queue for playback
    if (message.serverContent?.modelTurn?.parts) {
      for (const part of message.serverContent.modelTurn.parts) {
        if (part.inlineData?.data) {
          this.audioManager.queuePlayback(part.inlineData.data);
        }
      }
    }

    // Transcription — show in UI and log
    if (message.serverContent?.outputTranscription?.text) {
      const text = message.serverContent.outputTranscription.text;
      log("NARRATION", `Said: ${text}`);
      this.onTranscript(text);
    }
  }

  /**
   * Send a Claude activity event for narration.
   * Events are batched to avoid overwhelming the narrator.
   */
  sendEvent(description: string): void {
    if (!this.session || !this.connected || this.muted) return;

    this.eventBuffer.push(description);

    if (!this.flushTimer) {
      this.flushTimer = setTimeout(() => {
        this.flushEvents();
      }, this.FLUSH_DELAY_MS);
    }
  }

  /**
   * Flush all buffered events as a single message to the narrator.
   */
  private flushEvents(): void {
    this.flushTimer = null;
    if (!this.session || this.eventBuffer.length === 0 || this.muted) {
      this.eventBuffer = [];
      return;
    }

    const message = this.eventBuffer.join("\n");
    this.eventBuffer = [];

    try {
      this.session.sendClientContent({
        turns: [{ role: "user", parts: [{ text: message }] }],
        turnComplete: true,
      });
      log("NARRATION", `Sent update: ${message.slice(0, 120)}`);
    } catch (err) {
      log("NARRATION", `Error sending: ${err}`);
    }
  }

  /**
   * Send an immediate message (not batched). Use for important events
   * like "Claude started working" or "Task complete".
   */
  sendImmediate(text: string): void {
    if (!this.session || !this.connected || this.muted) return;

    // Flush any pending events first
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    if (this.eventBuffer.length > 0) {
      text = this.eventBuffer.join("\n") + "\n" + text;
      this.eventBuffer = [];
    }

    try {
      this.session.sendClientContent({
        turns: [{ role: "user", parts: [{ text }] }],
        turnComplete: true,
      });
      log("NARRATION", `Sent immediate: ${text.slice(0, 120)}`);
    } catch (err) {
      log("NARRATION", `Error sending: ${err}`);
    }
  }

  /**
   * Mute narration and clear any queued audio.
   * Call this before main Gemini is about to speak.
   */
  silence(): void {
    this.muted = true;
    this.eventBuffer = [];
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    this.audioManager.clearPlayback();
    log("NARRATION", "Silenced");
  }

  /**
   * Unmute narration. Call when main Gemini enters function call wait.
   */
  unmute(): void {
    this.muted = false;
    log("NARRATION", "Unmuted");
  }

  isConnected(): boolean {
    return this.connected;
  }

  async disconnect(): Promise<void> {
    this.muted = true;
    this.eventBuffer = [];
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    if (this.session) {
      try {
        this.session.close();
      } catch {
        // Already closed
      }
      this.session = null;
    }
    this.connected = false;
  }
}

File: VoicedVibe/Sources/VoicedVibe/Services/NarrationConnection.swift
import Foundation

protocol NarrationConnectionDelegate: AnyObject, Sendable {
    @MainActor func narrationDidReceiveAudio(base64: String)
    @MainActor func narrationDidReceiveTranscript(text: String)
    @MainActor func narrationTurnComplete()
}

actor NarrationConnection {
    private var webSocketTask: URLSessionWebSocketTask?
    private weak var delegate: NarrationConnectionDelegate?
    private var muted = false
    private var connected = false
    private var disconnecting = false
    private var languageCode: String
    private var backendBaseURL: String

    private var eventBuffer: [String] = []
    private var flushTask: Task<Void, Never>?
    private let flushDelay: Duration = .milliseconds(1500)

    init(delegate: NarrationConnectionDelegate, languageCode: String, backendBaseURL: String) {
        self.delegate = delegate
        self.languageCode = languageCode
        self.backendBaseURL = backendBaseURL
    }

    func connect() async {
        disconnecting = false

        do {
            let token = try await fetchToken()
            let config = try await fetchNarrationConfig()

            let wsURLString = "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1alpha.GenerativeService.BidiGenerateContent?key=\(token.token)"

            guard let url = URL(string: wsURLString) else { return }

            let session = URLSession(configuration: .default)
            let task = session.webSocketTask(with: url)
            webSocketTask = task
            task.resume()

            let setupMessage = buildSetupMessage(model: config.model, systemPrompt: config.system_prompt)
            try await sendJSON(setupMessage)

            receiveLoop()
            print("[NarrationConnection] Connected")
        } catch {
            print("[NarrationConnection] Connect failed: \(error)")
            if !disconnecting {
                scheduleReconnect()
            }
        }
    }

    func disconnect() {
        disconnecting = true
        muted = true
        eventBuffer.removeAll()
        flushTask?.cancel()
        flushTask = nil
        webSocketTask?.cancel(with: .goingAway, reason: nil)
        webSocketTask = nil
        connected = false
    }

    func sendEvent(_ description: String) {
        guard connected, !muted else { return }

        eventBuffer.append(description)

        if flushTask == nil {
            flushTask = Task { [weak self] in
                try? await Task.sleep(for: .milliseconds(1500))
                await self?.flushEvents()
            }
        }
    }

    func sendImmediate(_ text: String) {
        guard connected, !muted else { return }

        flushTask?.cancel()
        flushTask = nil

        var message = text
        if !eventBuffer.isEmpty {
            message = eventBuffer.joined(separator: "\n") + "\n" + text
            eventBuffer.removeAll()
        }

        let msg: [String: Any] = ["realtimeInput": ["text": message]]
        Task { try? await sendJSON(msg) }
    }

    func silence() {
        muted = true
        eventBuffer.removeAll()
        flushTask?.cancel()
        flushTask = nil
        Task { await delegate?.narrationDidReceiveAudio(base64: "") }
    }

    func unmute() {
        muted = false
    }

    // MARK: - Private

    private func buildSetupMessage(model: String, systemPrompt: String) -> [String: Any] {
        [
            "setup": [
                "model": "models/\(model)",
                "generationConfig": [
                    "responseModalities": ["AUDIO"],
                    "speechConfig": [
                        "languageCode": languageCode,
                        "voiceConfig": [
                            "prebuiltVoiceConfig": ["voiceName": "Algenib"],
                        ],
                    ] as [String: Any],
                    "thinkingConfig": [
                        "thinkingLevel": "HIGH",
                    ],
                ] as [String: Any],
                "systemInstruction": [
                    "parts": [["text": systemPrompt]],
                ],
                "realtimeInputConfig": [
                    "automaticActivityDetection": ["disabled": true],
                ],
                "contextWindowCompression": [
                    "triggerTokens": 104857,
                    "slidingWindow": ["targetTokens": 52428],
                ],
                "outputAudioTranscription": [:] as [String: Any],
            ] as [String: Any],
        ]
    }

    private func receiveLoop() {
        guard let task = webSocketTask else { return }

        task.receive { [weak self] result in
            guard let self else { return }

            Task {
                switch result {
                case .success(let message):
                    switch message {
                    case .string(let text):
                        await self.handleMessage(text)
                    case .data(let data):
                        if let text = String(data: data, encoding: .utf8) {
                            await self.handleMessage(text)
                        }
                    @unknown default:
                        break
                    }
                    await self.receiveLoop()

                case .failure(let error):
                    print("[NarrationConnection] Receive error: \(error)")
                    await self.handleDisconnect()
                }
            }
        }
    }

    private func handleMessage(_ text: String) async {
        guard let data = text.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return }

        if json["setupComplete"] != nil, !connected {
            connected = true
            print("[NarrationConnection] Setup complete")
        }

        guard !muted else { return }

        if let serverContent = json["serverContent"] as? [String: Any],
           let modelTurn = serverContent["modelTurn"] as? [String: Any],
           let parts = modelTurn["parts"] as? [[String: Any]] {
            for part in parts {
                if let inlineData = part["inlineData"] as? [String: Any],
                   let audioData = inlineData["data"] as? String {
                    await delegate?.narrationDidReceiveAudio(base64: audioData)
                }
            }
        }

        if let serverContent = json["serverContent"] as? [String: Any],
           let outputTranscription = serverContent["outputTranscription"] as? [String: Any],
           let transcriptText = outputTranscription["text"] as? String {
            await delegate?.narrationDidReceiveTranscript(text: transcriptText)
        }

        if let serverContent = json["serverContent"] as? [String: Any],
           serverContent["turnComplete"] != nil {
            await delegate?.narrationTurnComplete()
        }
    }

    private func handleDisconnect() async {
        connected = false
        webSocketTask = nil
        if !disconnecting {
            scheduleReconnect()
        }
    }

    private func scheduleReconnect() {
        Task {
            try? await Task.sleep(for: .seconds(5))
            guard !disconnecting else { return }
            await connect()
        }
    }

    private func flushEvents() {
        flushTask = nil
        guard !eventBuffer.isEmpty, !muted else {
            eventBuffer.removeAll()
            return
        }

        let message = eventBuffer.joined(separator: "\n")
        eventBuffer.removeAll()

        let msg: [String: Any] = ["realtimeInput": ["text": message]]
        Task { try? await sendJSON(msg) }
    }

    private func sendJSON(_ dict: [String: Any]) async throws {
        guard let task = webSocketTask else { return }
        let data = try JSONSerialization.data(withJSONObject: dict)
        guard let string = String(data: data, encoding: .utf8) else { return }
        try await task.send(.string(string))
    }

    private func fetchToken() async throws -> TokenResponse {
        let url = URL(string: "\(backendBaseURL)/api/token")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(TokenResponse.self, from: data)
    }

    private func fetchNarrationConfig() async throws -> ServerConfig {
        let url = URL(string: "\(backendBaseURL)/api/narration-config")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return try JSONDecoder().decode(ServerConfig.self, from: data)
    }
}
